_posts/2018-11-29-GraphQL-Stitching.markdown

```diff
@@ -0,0 +1,386 @@
+2017][stitching_out] via the [`graphql-tools`][tools] and became production-[ready in April
+2018][stitching_announcement].
+
+We started experimenting on staging last year and would occasionally run into edge-case issues.This meant the state
+of the project would ebb & flow between being blocked, or no-one having the bandwidth to work on it. This was fine,
+because our aim was [incremental evolutions over bold revolution][rev].
+
+Before we dive into implementation details, here's a quick glossary of terms before we start:
+
+- **GraphQL Type** - the shape of an object exposed from your GraphQL API
+- **GraphQL Schema** - a representation of your GraphQL's type system, containing all types and fields on them
+- **GraphQL Resolver** - every field accessed in a query resolves to a corresponding value, the function doing that
+  is a resolver
+- **Schema Merging** - taking two GraphQL schemas, and merging all the types and resolvers into one schema
+- **Schema Stitching** - extending a GraphQL Schema programmatically, with the ability to delegate to merged
+  schemas
+
+Stitching is one of the end-goals, but merging may be enough for a lot of cases. Both of the two launch posts above
+give a much more in-depth explanation of how everything comes together, but these should be enough for this post.
+
+## How Do We Do It?
+
+We have 5 GraphQL APIs inside the Artsy ecosystem, our aim is to cautiously include these APIs inside metaphysics.
+We don't need the entire contents of those APIs, and as you'll learn - we couldn't do that even if we wanted.
+
+The technique we settled on was:
+
+1.  Download the schema of each external API into metaphysics' source code
+1.  Have each schema trimmed to just the essentials that we need today
+1.  Merge in each schema incrementally
+1.  Stitch in any desired schema changes
+
+Let's dig, with some code into how we do each of these steps.
+
+#### Downloading Schemas
+
+We created a [pretty minimal script][dl-schema] which can be run periodically from a developer's computer.
+
+```js
+const destination = "src/data"
+
+const httpConvectionLink = createHttpLink({
+  fetch,
+  uri: urljoin("https://convection-staging.artsy.net/api", "graphql")
+})
+
+introspectSchema(httpConvectionLink).then(schema => {
+  fs.writeFileSync(
+    path.join(destination, "convection.graphql"),
+    printSchema(schema, { commentDescriptions: true })
+  )
+})
+```
+
+The script uses an [apollo-http-link][] to grab our schema, and store it in our repo, see
+[`src/data/convection.graphql`][c-gql]. This means that when someone wants to update to a new version of the
+schema, it will go through code review and a normal testing-flow. The trade-off being that it will always be out of
+date a little bit, but you can make guarantees about the current schema. This is a reasonable trade-off, as GraphQL
+schemas should always be forward compatible for queries, and when someone wants to use a new field from another
```

Comment:
> should always be forward compatible for queries

I think we have some automated (Peril?) code to ensure this, might be worth linking to.